### ISSUE TITLE
Support for Automatic Detection and Utilization of GOCACHEPROG Environment Variable

### ISSUE THREAD

**Issue by ryancurrah** [goreleaser/goreleaser#4939]
### Is your feature request related to a problem? Please describe.

I propose that GoReleaser should automatically detect the `GOCACHEPROG` environment variable and pass it to child processes to ensure consistent usage of user-defined cache programs across builds.

As detailed in [golang/go#59719](https://github.com/golang/go/issues/59719), Go has introduced a new environment variable `GOCACHEPROG` that allows users to specify a custom program for caching mechanisms. This program is designed to run as a child process, where `cmd/go` communicates with it via stdin/stdout. This setup translates the Go tool's internal cache interface, allowing the `GOCACHEPROG` to implement any desired caching mechanism or policy.

The primary concern is that if GoReleaser does not automatically pass this variable to child processes, users might incorrectly assume their builds are utilizing the defined remote cache, leading to potential inconsistencies and issues in build environments where caching is crucial.

### Describe the solution you'd like

GoReleaser should check for the `GOCACHEPROG` environment variable during its execution. If found, GoReleaser should ensure that this variable is effectively used in all child processes that are spawned during the build process. This behavior will help maintain consistency across builds and ensure that all components are correctly utilizing the intended caching mechanisms.

### Describe alternatives you've considered

One alternative could involve manually configuring the `Env` section of the GoReleaser configuration to pass `GOCACHEPROG`. However, this is not ideal for a few reasons:
- **User Expectations:** Users often set such environment variables globally, especially in CI environments, expecting them to be automatically recognized and utilized by all tools in the toolchain.
- **Consistency and Usability:** Requiring manual configuration adds an unnecessary layer of complexity and could lead to errors or inconsistencies if the variable is overlooked or misconfigured.

### Search

- [X] I did search for other open and closed issues before opening this

### Supporter

- [X] I am a [sponsor](https://github.com/sponsors/caarlos0/) or a [Pro](https://goreleaser.com/pro) customer

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

Implementing this feature would enhance GoReleaser's functionality by supporting newer features of the Go toolchain and improving compatibility with custom caching solutions, ultimately making it more robust and user-friendly for diverse Go CI environments.

Thank you for considering this feature request. I am looking forward to your feedback and am happy to contribute to the discussion or implementation.

**Comment by caarlos0** [comment]
hey @ryancurrah 

This is already possible:


```yaml
# ...
builds:
  - # ...
    env:
      - "GOCACHEPROG={{.Env.GOCACHEPROG}}"
# ...
```

and then:

```bash
GOCACHEPROG=myprog goreleaser 
```


Let me know if that solves your issue!

**Comment by ryancurrah** [comment]
It's not really feasible to do that for 800 Go modules. Was hoping it would just automatically look for it.

**Comment by caarlos0** [comment]
it seems it also requires `GOEXPERIMENT=cacheprog`?

if so, we would need to set that one too, right?

**Comment by ryancurrah** [comment]
No thats only used when compiling Go to support it.

**Comment by caarlos0** [comment]
something like this then: https://github.com/goreleaser/goreleaser/pull/4941

**Comment by ryancurrah** [comment]
Looks good to me.

**Comment by caarlos0** [comment]
thanks!

fixed in #4941
